cranelift/codegen/src/isa/x64/lower.rs

```diff
@@ -896,48 +896,106 @@ fn lower_insn_to_regs<C: LowerCtx<I = Inst>>(
+                    Opcode::Fadd => SseOpcode::Addps,
+                    Opcode::Fsub => SseOpcode::Subps,
+                    Opcode::Fmul => SseOpcode::Mulps,
+                    Opcode::Fdiv => SseOpcode::Divps,
+                    _ => unreachable!(),
+                },
+                types::F64X2 => match op {
+                    Opcode::Fadd => SseOpcode::Addpd,
+                    Opcode::Fsub => SseOpcode::Subpd,
+                    Opcode::Fmul => SseOpcode::Mulpd,
+                    Opcode::Fdiv => SseOpcode::Divpd,
+                    _ => unreachable!(),
+                },
+                _ => panic!(
+                    "invalid type: expected one of [F32, F64, F32X4, F64X2], found {}",
+                    ty
+                ),
             };
+            ctx.emit(Inst::xmm_rm_r(sse_op, rhs, dst));
+        }
 
-            let is_64 = flt_ty_is_64(ty.unwrap());
+        Opcode::Fneg => {
+            let src = input_to_reg_mem(ctx, inputs[0]);
+            let dst = output_to_reg(ctx, outputs[0]);
+            let ty = ty.unwrap();
+
+            // Move the `lhs` to the same register as `dst`; this may not emit an actual move
+            // but ensures that the registers are the same to match x86's read-write operand
+            // encoding.
+            match src {
+                RegMem::Reg { reg } => ctx.emit(Inst::gen_move(dst, reg, ty)),
+                RegMem::Mem { .. } => unimplemented!(),
+            }
 
-            let mov_op = if is_64 {
-                SseOpcode::Movsd
-            } else {
-                SseOpcode::Movss
-            };
-            ctx.emit(Inst::xmm_mov(mov_op, lhs, dst, None));
+            // Generate an all 1s constant in an XMM register.
+            let tmp = ctx.alloc_tmp(RegClass::V128, ty);
+            let cond: FloatComparisonPredicate = FloatCC::Equal.into();
+            let cmpps = Inst::xmm_rm_r_imm(
+                SseOpcode::Cmpps,
+                RegMem::reg(tmp.to_reg()),
+                tmp,
+                cond.encode(),
+            );
+            ctx.emit(cmpps);
 
-            let sse_op = if is_64 { f64_op } else { f32_op };
-            ctx.emit(Inst::xmm_rm_r(sse_op, RegMem::reg(rhs), dst));
+            // Shift it so only the first bit is set.
+            let shift_by = RegMemImm::imm((ty.lane_bits() - 1) as u32);
+            let shift = Inst::xmm_shift(ShiftKind::ShiftLeft, LaneWidth::from(ty), shift_by, tmp);
+            ctx.emit(shift);
+
+            // Invert the sign bit.
+            let xor = Inst::xmm_rm_r(SseOpcode::Xorps, RegMem::reg(tmp.to_reg()), dst);
+            ctx.emit(xor);
         }
 
         Opcode::Sqrt => {
             let src = input_to_reg_mem(ctx, inputs[0]);
             let dst = output_to_reg(ctx, outputs[0]);
-
-            let (f32_op, f64_op) = match op {
-                Opcode::Sqrt => (SseOpcode::Sqrtss, SseOpcode::Sqrtsd),
-                _ => unreachable!(),
+            let ty = ty.unwrap();
+
```

Comment:
Should there be a `gen_move` here as well (for `Opcode::Sqrt`)?

Reply:
I don't think so, since sqrt doesn't have any read-write operand. Am i missing something?

Reply:
I agree with @bnjbvr; the insn reads its input register and writes a different output register, but there's no `mod`-ify style behaviour.  Note that `get_regs` for `XMM_RegMem_Reg` will have to take this into account; that it, it will have to special-case `SseOpcode::Sqrtss` et al accordingly.

Reply:
@julian-seward1 Sqrtss uses a different Vcode inst because of this, since it's rather an "unary" opcode (see also `XmmUnaryRmR`).